### PR TITLE
IGNITE-20708 Support cacheGroupNames for DumpReader

### PR DESCRIPTION
**Pull request by yurinaryshkin**
Thank you for submitting the pull request to the Apache Ignite.

In order to streamline the review of the contribution 
we ask you to ensure the following steps have been taken:

### The Contribution Checklist
- [ ] There is a single JIRA ticket related to the pull request. 
- [ ] The web-link to the pull request is attached to the JIRA ticket.
- [ ] The JIRA ticket has the _Patch Available_ state.
- [ ] The pull request body describes changes that have been made. 
The description explains _WHAT_ and _WHY_ was made instead of _HOW_.
- [ ] The pull request title is treated as the final commit message. 
The following pattern must be used: `IGNITE-XXXX Change summary` where `XXXX` - number of JIRA issue.
- [ ] A reviewer has been mentioned through the JIRA comments 
(see [the Maintainers list](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute#HowtoContribute-ReviewProcessandMaintainers)) 
- [ ] The pull request has been checked by the Teamcity Bot and 
the `green visa` attached to the JIRA ticket (see [TC.Bot: Check PR](https://mtcga.gridgain.com/prs.html))

### Notes
- [How to Contribute](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute)
- [Coding abbreviation rules](https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules)
- [Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines)
- [Apache Ignite Teamcity Bot](https://cwiki.apache.org/confluence/display/IGNITE/Apache+Ignite+Teamcity+Bot)

If you need any help, please email dev@ignite.apache.org or ask anу advice on http://asf.slack.com _#ignite_ channel.

### CODE REVIEW

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/dump/AbstractCacheDumpTest.java [hunk 1]
```diff
@@ -351,8 +360,8 @@ else if (data.config().getName().equals(CACHE_1)) {
                         throw new IgniteException("Unknown cache");
                 });
 
-                for (boolean found : cachesFound)
-                    assertTrue(found);
+                for (int i = 0; i < cachesFound.length; i++)
```
Comment (nizhikov) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/dump/AbstractCacheDumpTest.java:363]: cachesFound sorted? Do we have guarantee that order of caches in cachesFound will be the same as we expected?

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/dump/AbstractCacheDumpTest.java [hunk 2]
```diff
@@ -284,6 +284,15 @@ protected void checkDump(IgniteEx ign) throws Exception {
 
     /** */
     void checkDump(IgniteEx ign, String name) throws Exception {
+        checkDump(ign, name, null, new boolean[]{true, true, true},
+            KEYS_CNT + (onlyPrimary ? 0 : KEYS_CNT * backups),
+            2 * (KEYS_CNT + (onlyPrimary ? 0 : KEYS_CNT * backups)), KEYS_CNT);
+    }
+
+    /** */
+    void checkDump(IgniteEx ign, String name, String[] cacheGroupNames, boolean[] expectedFoundCaches,
```
Comment (nizhikov) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/dump/AbstractCacheDumpTest.java:293]: Please, use parameter formatting from coding style.

Снимок экрана 2023-10-20 в 17 09 40

Comment (nizhikov) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/dump/AbstractCacheDumpTest.java:293]: Let's rewrite this as `Set<String> expectedFoundCaches` with cache names we are expecting to read from dump.

Comment (yurinaryshkin) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/dump/AbstractCacheDumpTest.java]: Rewrote to use Set.

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/dump/AbstractCacheDumpTest.java (hunk 1 above)
Comment (yurinaryshkin) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/dump/AbstractCacheDumpTest.java]: Updated the logic to compare 2 Sets.

File: modules/core/src/main/java/org/apache/ignite/dump/DumpReader.java [hunk 3]
```diff
@@ -85,9 +89,14 @@ public DumpReader(DumpReaderConfiguration cfg, IgniteLogger log) {
 
                 Map<Integer, List<String>> grpToNodes = new HashMap<>();
 
+                Set<Integer> cacheGroupIds = cfg.cacheGroupNames() == null ? null :
+                    Arrays.stream(cfg.cacheGroupNames()).map(CU::cacheId).collect(Collectors.toSet());
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/dump/DumpReader.java:93]:
```suggestion
                Set<Integer> cacheGroupIds = cfg.cacheGroupNames() != null 
                    ? Arrays.stream(cfg.cacheGroupNames()).map(CU::cacheId).collect(Collectors.toSet())
                    : null;
```

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/dump/AbstractCacheDumpTest.java [hunk 4]
```diff
@@ -284,6 +285,24 @@ protected void checkDump(IgniteEx ign) throws Exception {
...
+            null,
+            new HashSet<>(Arrays.asList(DEFAULT_CACHE_NAME, CACHE_0, CACHE_1)),
+            KEYS_CNT + (onlyPrimary ? 0 : KEYS_CNT * backups),
+            2 * (KEYS_CNT + (onlyPrimary ? 0 : KEYS_CNT * backups)),
+            KEYS_CNT);
+    }
+
+    /** */
+    void checkDump(IgniteEx ign,
+        String name,
```
Comment (nizhikov) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/dump/AbstractCacheDumpTest.java:299]:
```suggestion
    void checkDump(
        IgniteEx ign,
        String name,
```

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/dump/IgniteCacheDumpSelfTest.java [hunk 5]
```diff
@@ -146,6 +146,45 @@ public void testCacheDump() throws Exception {
...
+    @Test
+    public void testCacheDumpWithReadGroupFilter() throws Exception {
+        snpPoolSz = 4;
+
+        try {
+            IgniteEx ign = startGridAndFillCaches();
+
+            createDump(ign);
+
+            checkDump(ign,
```
Comment (nizhikov) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/dump/IgniteCacheDumpSelfTest.java:159]:
```suggestion
            checkDump(
                ign,
```

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/dump/IgniteCacheDumpSelfTest.java [hunk 6]
```diff
@@ -146,6 +146,45 @@ public void testCacheDump() throws Exception {
...
+                DMP_NAME,
+                new String[]{GRP},
+                new HashSet<>(Arrays.asList(CACHE_0, CACHE_1)),
+                0,
+                2 * (KEYS_CNT + (onlyPrimary ? 0 : KEYS_CNT * backups)),
+                0);
+
+            checkDump(ign,
```
Comment (nizhikov) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/dump/IgniteCacheDumpSelfTest.java:167]:
```suggestion
            checkDump(
                ign,
```

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/dump/IgniteCacheDumpSelfTest.java [hunk 7]
```diff
@@ -146,6 +146,45 @@ public void testCacheDump() throws Exception {
...
+                0,
+                KEYS_CNT);
+
+            checkDump(ign,
+                DMP_NAME,
+                new String[]{DEFAULT_CACHE_NAME, GRP},
+                new HashSet<>(Arrays.asList(DEFAULT_CACHE_NAME, CACHE_0, CACHE_1)),
+                KEYS_CNT + (onlyPrimary ? 0 : KEYS_CNT * backups),
+                2 * (KEYS_CNT + (onlyPrimary ? 0 : KEYS_CNT * backups)),
+                KEYS_CNT);
```
Comment (nizhikov) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/dump/IgniteCacheDumpSelfTest.java:181]:
```suggestion
            checkDump(
                ign,
                DMP_NAME,
                new String[]{DEFAULT_CACHE_NAME, GRP},
                new HashSet<>(Arrays.asList(DEFAULT_CACHE_NAME, CACHE_0, CACHE_1)),
                KEYS_CNT + (onlyPrimary ? 0 : KEYS_CNT * backups),
                2 * (KEYS_CNT + (onlyPrimary ? 0 : KEYS_CNT * backups)),
                KEYS_CNT
            );
```

**Review by nizhikov — APPROVED** [review]

**Comment by nizhikov** [comment]
Left a minor suggestions about codestyle.
Please, provide bot visa in ticket to be sure these changes don't raise any regression.

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/dump/IgniteCacheDumpSelfTest.java [hunk 8]
```diff
@@ -146,6 +146,50 @@ public void testCacheDump() throws Exception {
...
+            createDump(ign);
+
+            checkDump(
+                ign,
+                DMP_NAME,
+                new String[]{GRP},
+                new HashSet<>(Arrays.asList(CACHE_0, CACHE_1)),
+                0,
+                2 * (KEYS_CNT + (onlyPrimary ? 0 : KEYS_CNT * backups)),
+                0);
```
Comment (nizhikov) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/dump/IgniteCacheDumpSelfTest.java:166]:
```suggestion
                0
            );
```

**Pull request merged**